glue-codes/openfast-cpp/src/OpenFAST.cpp

```diff
@@ -7,304 +8,722 @@ int fast::OpenFAST::AbortErrLev = ErrID_Fatal; // abort error level; compare wit
+        if(scStatus) {
+            
+            sc->init(nTurbinesGlob, numScInputs, numScOutputs);
+            
+            sc->calcOutputs(scOutputsGlob);
+            fillScOutputsLoc();
+        }
+        
+        // Unfortunately setVelocity only sets the velocity at 'n+1'. Need to copy 'n+1' to 'n'
+        init_velForceNodeData() ;
+        send_data_to_openfast(fast::np1);
+        
+        for (int iTurb=0; iTurb < nTurbinesProc; iTurb++) {
+            
+            FAST_CFD_Solution0(&iTurb, &ErrStat, ErrMsg);
+            checkError(ErrStat, ErrMsg);
+            
+            FAST_CFD_InitIOarrays_SS(&iTurb, &ErrStat, ErrMsg);
+            checkError(ErrStat, ErrMsg);       
+        }
+        
+        get_data_from_openfast(fast::n);
+        get_data_from_openfast(fast::nm1);
+        get_data_from_openfast(fast::nm2);        
+        
+        timeZero = false;
 
-  if (!dryRun) {
-       // set wind speeds at initial locations
-    // for (int iTurb=0; iTurb < nTurbinesProc; iTurb++) {
-    //     setOutputsToFAST(cDriver_Input_from_FAST[iTurb], cDriver_Output_to_FAST[iTurb]);
-    // }
-     
-     if(scStatus) {
+        if (scStatus) {
+            fillScInputsGlob(); // Update inputs to super controller
+        }
+    }
+    
+}
 
-       sc->init(nTurbinesGlob, numScInputs, numScOutputs);
+void fast::OpenFAST::init_velForceNodeData() {
 
-       sc->calcOutputs(scOutputsGlob);
-       fillScOutputsLoc();
-     }
+    for (int iTurb=0; iTurb < nTurbinesProc; iTurb++) {
+        int nvelpts = get_numVelPtsLoc(iTurb);
+        int nfpts = get_numForcePtsLoc(iTurb);
+        for (int i=0; i<nvelpts; i++) {
+            for (int j=0 ; j < 3; j++) {
+                velForceNodeData[iTurb][fast::n].x_vel[i*3+j] = velForceNodeData[iTurb][fast::np1].x_vel[i*3+j];
+                velForceNodeData[iTurb][fast::nm1].x_vel[i*3+j] = velForceNodeData[iTurb][fast::np1].x_vel[i*3+j];
+                velForceNodeData[iTurb][fast::nm2].x_vel[i*3+j] = velForceNodeData[iTurb][fast::np1].x_vel[i*3+j];
+                velForceNodeData[iTurb][fast::n].xdot_vel[i*3+j] = velForceNodeData[iTurb][fast::np1].xdot_vel[i*3+j];
+                velForceNodeData[iTurb][fast::nm1].xdot_vel[i*3+j] = velForceNodeData[iTurb][fast::np1].xdot_vel[i*3+j];
+                velForceNodeData[iTurb][fast::nm2].xdot_vel[i*3+j] = velForceNodeData[iTurb][fast::np1].xdot_vel[i*3+j];
+                velForceNodeData[iTurb][fast::n].vel_vel[i*3+j] = velForceNodeData[iTurb][fast::np1].vel_vel[i*3+j];
+                velForceNodeData[iTurb][fast::nm1].vel_vel[i*3+j] = velForceNodeData[iTurb][fast::np1].vel_vel[i*3+j];
+                velForceNodeData[iTurb][fast::nm2].vel_vel[i*3+j] = velForceNodeData[iTurb][fast::np1].vel_vel[i*3+j];
+            }
+        }
+        for (int i=0; i<nfpts; i++) {
+            for (int j=0 ; j < 3; j++) {
+                velForceNodeData[iTurb][fast::n].x_force[i*3+j] = velForceNodeData[iTurb][fast::np1].x_force[i*3+j];
```

Comment:
Recommend creating some `set_state`, `swap_states` functions.

Reply:
Thanks. This comment was super useful. Reduced the code quite a bit! Implemented in 9f34f9eb119d38dc0c78bfe58fdd23d084d02430.